src/Datadog.Trace/HttpOverStreams/DatadogHttpClient.cs

```diff
@@ -0,0 +1,188 @@
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Datadog.Trace.HttpOverStreams.HttpContent;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.HttpOverStreams
+{
+    internal class DatadogHttpClient
+    {
+        /// <summary>
+        /// Typical headers sent to the agent are small.
+        /// Allow enough room for future expansion of headers.
+        /// </summary>
+        public const int MaxRequestHeadersBufferSize = 2560;
+
+        private const string ContentLengthHeaderKey = "Content-Length";
+
+        private static readonly Vendors.Serilog.ILogger Logger = DatadogLogging.For<DatadogHttpClient>();
+
+        public async Task<HttpResponse> SendAsync(HttpRequest request, Stream requestStream, Stream responseStream)
+        {
+            await SendRequestAsync(request, requestStream).ConfigureAwait(false);
+            return ReadResponse(responseStream);
+        }
+
+        private async Task SendRequestAsync(HttpRequest request, Stream requestStream)
+        {
+            // Headers are always ASCII per the HTTP spec
+            using (var writer = new StreamWriter(requestStream, Encoding.ASCII, MaxRequestHeadersBufferSize, leaveOpen: true))
+            {
+                await DatadogHttpHeaderHelper.WriteLeadingHeaders(request, writer).ConfigureAwait(false);
+
+                foreach (var header in request.Headers)
+                {
+                    await DatadogHttpHeaderHelper.WriteHeader(writer, header).ConfigureAwait(false);
+                }
+
+                await DatadogHttpHeaderHelper.WriteEndOfHeaders(writer).ConfigureAwait(false);
+            }
+
+            await request.Content.CopyToAsync(requestStream, DatadogHttpValues.MaximumRequestBufferSize).ConfigureAwait(false);
+            Logger.Debug("Datadog HTTP: Flushing stream.");
+            await requestStream.FlushAsync().ConfigureAwait(false);
+        }
+
+        private HttpResponse ReadResponse(Stream responseStream)
+        {
+            var headers = new HttpHeaders();
+            char currentChar;
+            int streamPosition = 0;
+
+            // https://tools.ietf.org/html/rfc2616#section-4.2
+            // HTTP/1.1 200 OK
+            // HTTP/1.1 XXX MESSAGE
+
+            const int statusCodeStart = 9;
+            const int statusCodeEnd = 12;
+            const int startOfReasonPhrase = 13;
+
+            // TODO: Get this from StringBuilderCache after we determine safe maximum capacity
+            var stringBuilder = new StringBuilder();
+
+            var chArray = new byte[1];
+            void GoNextChar()
+            {
+                streamPosition++;
+                chArray[0] = (byte)responseStream.ReadByte();
```

Comment:
what a great place to use stackalloc... :'(

Reply:
Is weird to me that this local function uses a parent variable for output. Why not using char as return value?

Reply:
Overall I think this ReadResponse method can be revisit later for optimizations. Don't know if there is a better way than reading char by char from the stream and calling Encoding.ASCII.GetChars for each char. It looks expensive.

Reply:
Would it help to store the byte value for each of the chars that we use as indicators?